### PR TITLE
Classifier refactor

### PR DESCRIPTION
**Pull request by brian-c**
This better isolates classifier components' responsibilities.

The Classifier now keeps track of the state of the current classification and task/annotation. Tasks can implement `onLeaveAnnotation` to react to moving on to the next annotation (e.g. The drawing task forcing any incomplete shapes to complete).

The subject annotator displays the subject and provides hooks for individual tasks to modify it however they need to. Tasks can implement `InsideSubject` (e.g. drawing task's mark initializer), `PersistInsideSubject` (the drawing task's marks renderer, which persists even when you're not on a drawing task), and `AfterSubject` (current survey choices) components to modify the subject annotator. We can easily add more hooks as we require them.

Drawing marks' sub-task details popup are now rendered by the drawing tool root using the new modal-form module, which is way, way saner than hacking the right position the tether/Tooltip module. This should be the only visible change.

This also sets things up for required drawing tasks (#488) and sub-tasks (#251) and drawing tools handling their own `forceComplete` behavior (#1310).

Big changes. Please ask if something seems weird.

### CODE REVIEW

File: app/classifier/index.cjsx [hunk 1]
```diff
@@ -41,24 +39,30 @@ Classifier = React.createClass
...
     @setState subjectLoading: true
-    # TODO: Pre-cache subject images here and change loading state back.
+    preloadSubject subject
+      .then =>
+        if @props.subject is subject # The subject could have changed while we were loading.
+          @setState subjectLoading: false
+          @props.onLoad?()
 
   prepareToClassify: (classification) ->
     setTimeout => # TODO: Why is this here?
```
Comment (aweiksnar) [inline app/classifier/index.cjsx]: Is this setTimeout still needed? I've been consistently getting a defined `classification` without it

Comment (brian-c) [inline app/classifier/index.cjsx]: Looks like it's no longer needed. I've removed it.

**Comment by brian-c** [comment]
Just rebased and force-pushed.

**Comment by aweiksnar** [comment]
Just noticed something: It seems that some marks are now appearing offset to the cursor. This is reproducible if you go into the fossil-finder project and make some marks in the first step

**Comment by brian-c** [comment]
Boo. Will investigate, thanks. I'm noticing some subject resizing issues now too. Put this on hold for a bit.

**Comment by aweiksnar** [comment]
Cool will do

**Comment by brian-c** [comment]
I think I got it: To reproduce consistently, go to the classify page, scroll down, go to another page, and hit the browser's Back button. That'll mount the classify page with its initial `state.sizeRect` offset by the scroll amount.

Fixed by factoring in current scroll when setting the sizeRect.

**Comment by aweiksnar** [comment]
:+1:

**Pull request merged**